Source/WebCore/html/HTMLCanvasElement.cpp

```diff
@@ -538,6 +555,49 @@ ImageBitmapRenderingContext* HTMLCanvasElement::getContextBitmapRenderer(const S
     return static_cast<ImageBitmapRenderingContext*>(m_context.get());
 }
 
+bool HTMLCanvasElement::isWebGPUType(const String& type)
+{
+    return type == "webgpu"_s;
+}
+
+GPUCanvasContext* HTMLCanvasElement::createContextWebGPU(const String& type)
+{
+    ASSERT_UNUSED(type, HTMLCanvasElement::isWebGPUType(type));
+    ASSERT(!m_context);
+
+    if (!document().settings().webGPU())
+        return nullptr;
+
+    m_context = GPUCanvasContext::create(*this);
+
+    if (m_context) {
+        // Adds the current document as an observer, so that the document calls prepareForDisplay
+        // when the canvas changes.
+        addObserver(document());
+
+        // Need to make sure a RenderLayer and compositing layer get created for the Canvas.
+        invalidateStyleAndLayerComposition();
+    }
+
+    return static_cast<GPUCanvasContext*>(m_context.get());
```

Comment:
Do you need the null check? and if so, the `m_context.get()` line will fail since `m_context` has the opportunity to be null at this point.

Reply:
Probably, we don't want to do unnecessary work if we can't create a context. WebGL does similarly.

Reply:
Ok but `GPUCanvasContext::create` never returns null unlike the WebGL sites.
```
std::unique_ptr<GPUCanvasContext> GPUCanvasContext::create(CanvasBase& canvas)
 {
     return std::unique_ptr<GPUCanvasContext>(new GPUCanvasContext(canvas));
 }
 ```

Reply:
Hmm okay, it's possible to create a non-functional `GPUCanvasContext` because it's running on platforms that don't support creating `WebGPUSurface` yet. I'll see if we can fix it so that it will refuse to create a GPUCanvasContext on those platforms.